### PR TITLE
support master_lb_enabled

### PR DESCRIPTION
**Pull request by xinnong-wang**
#125: Make Load Balancer for Kubernetes Master(s) an optional parameter
#137: Assign private IPs for master instances during instance creation

Approach to use a software LBR (Nginx) in place of the public oci master LBR.

**create the following nginx.conf**
```
error_log stderr notice;

worker_processes auto;
events {
  multi_accept on;
  use epoll;
  worker_connections 1024;
}

stream {
    upstream kube_apiserver {
        least_conn;
        server master1:443;
        server master2:443;
        server master3:443;
    }

    server {
        listen        0.0.0.0:6443;
        proxy_pass    kube_apiserver;
        proxy_timeout 10m;
        proxy_connect_timeout 1s;
    }
}
```
**start nginx**
sudo docker run -d 6443:6443 -v /etc/nginx:/etc/nginx --name nginx-proxy --restart=always nginx:1.13.1

**change kubeconfig to api-server: https://127.0.0.1:6443**
 kube-proxy.yaml and worker-kubeconfig.yaml on worker instances.
```
apiVersion: v1
kind: Config
clusters:
- name: local
  cluster:
    server: https://127.0.0.1:6443
```

### CODE REVIEW

File: instances/etcd/variables.tf [hunk 1]
```diff
@@ -1,28 +1,10 @@
 variable "network_cidrs" {
   type = "map"
-  default = {
```
Comment (jlamillan) [inline instances/etcd/variables.tf]: Why did these get deleted?

File: network/loadbalancers/k8smaster/variables.tf [hunk 2]
```diff
@@ -6,6 +6,10 @@ variable "label_prefix" {
   default = ""
 }
 
+variable "master_lb_enabled" {
+  default = true
+}
+
```
Comment (jlamillan) [inline network/loadbalancers/k8smaster/variables.tf]: I think we should stick with the convention of using the string values "true" and "false" for booleans. This would also be inline with your doc updates.

The version of terraform I am using at least is requiring me to set `master_lb_enabled = 1` or `master_lb_enabled = 0` as opposed to `master_lb_enabled = "true"` or `master_lb_enabled = "false"` to avoid a parsing trying to convert "false" to an integer in the `count`.

File: k8s-oci.tf [hunk 3]
```diff
@@ -471,9 +471,14 @@ module "k8smaster-public-lb" {
...
+  source = "./network/loadbalancers/nginx"
+  hosts  = "${concat(module.instances-k8smaster-ad1.private_ips,module.instances-k8smaster-ad2.private_ips, module.instances-k8smaster-ad3.private_ips)}"
+}
+
 module "kubeconfig" {
   source                     = "./kubernetes/kubeconfig"
   api_server_private_key_pem = "${module.k8s-tls.api_server_private_key_pem}"
   api_server_cert_pem        = "${module.k8s-tls.api_server_cert_pem}"
-  k8s_master                 = "https://${module.k8smaster-public-lb.ip_addresses[0]}:443"
+  k8s_master                 = "{local.master_address}"
```
Comment (jlamillan) [inline k8s-oci.tf]: Typo... should be `"${local.master_address}"`

File: network/loadbalancers/nginx/variable.tf [hunk 4]
```diff
@@ -0,0 +1,11 @@
+variable "version" {
+  default = "1.13.1"
```
Comment (jlamillan) [inline network/loadbalancers/nginx/variable.tf]: The nginx version should be an input variable.

File: docs/input-variables.md [hunk 5]
```diff
@@ -21,6 +21,7 @@ name                                | default                 | description
 etcdShape                           | VM.Standard1.1          | OCI shape for etcd nodes
 k8sMasterShape                      | VM.Standard1.1          | OCI shape for k8s master(s)
 k8sWorkerShape                      | VM.Standard1.2          | OCI shape for k8s worker(s)
+master_lb_enabled                   | true                    | enable/disable the k8s master load balancer. true: use the k8s master load balancer ip. false: use a nginx reverse proxy for k8s masters.
```
Comment (jlamillan) [inline docs/input-variables.md]: Since `false` _still_ gets you an LB (nginx) , I wonder if this input variable should be renamed to something like `master_oci_lb_enabled` to be a bit more descriptive.

File: instances/etcd/variables.tf (hunk 1 above)
Comment (jlamillan) [inline instances/etcd/variables.tf]: Why was this removed?

File: instances/etcd/variables.tf [hunk 6]
```diff
@@ -1,28 +1,10 @@
 variable "network_cidrs" {
   type = "map"
```
Comment (jlamillan) [inline instances/etcd/variables.tf]: Why was this removed?

File: network/loadbalancers/k8smaster/variables.tf (hunk 2 above)
Comment (xinnong-wang) [reply network/loadbalancers/k8smaster/variables.tf]: `master_lb_enabled` is defined as a boolean type (same as `etcd_iscsi_volume_create`), so you can set master_lb_enabled = `true` | `false` instead of 1 or 0.

Comment (xinnong-wang) [reply network/loadbalancers/k8smaster/variables.tf]: `etcd_lb_enabled` is defined as a `string`, in fact, it's better to change to a `boolean` also.
See https://www.terraform.io/docs/configuration/variables.html#booleans

File: k8s-oci.tf (hunk 3 above)
Comment (xinnong-wang) [reply k8s-oci.tf]: It can only be configured to the address of master, since it is the address of k8s api-server which exits only in master.  We can not set it to `worker_public_ip`.

When master_lb_enabled = false, api-server can't be accessed outside the cluster, so we have to run a nginx locally in NAT hosts which reverse proxy to the private addresses of masters.

File: instances/etcd/variables.tf (hunk 6 above)
Comment (xinnong-wang) [reply instances/etcd/variables.tf]: Because it is defined in <root>/variables.tf as a input variable, and should be passed here.
```
network_cidrs                       | See map in variables.tf | A CIDR notation IP range of the VCN and its subnets.
```

`network_cidrs` is shared by `instances-etcd` and `instances-k8smaster`.

**Comment by xinnong-wang** [comment]
@jlamillan Thanks for your comments.
Updated the scripts according your comments, could you please help review again?

File: docs/input-variables.md (hunk 5 above)
Comment (xinnong-wang) [reply docs/input-variables.md]: done

File: network/loadbalancers/nginx/variable.tf (hunk 4 above)
Comment (xinnong-wang) [reply network/loadbalancers/nginx/variable.tf]: done

File: instances/k8smaster/main.tf [hunk 7]
```diff
@@ -10,7 +10,14 @@ resource "oci_core_instance" "TFInstanceK8sMaster" {
...
   image               = "${lookup(data.oci_core_images.ImageOCID.images[0], "id")}"
   shape               = "${var.shape}"
-  subnet_id           = "${var.subnet_id}"
+
+  create_vnic_details {
+    subnet_id        = "${var.subnet_id}"
+    display_name     = "${var.label_prefix}${var.display_name_prefix}-${count.index}"
+    hostname_label   = "${var.hostname_label_prefix}-${count.index}"
+    assign_public_ip = "${(var.control_plane_subnet_access == "private") ? "false" : "true"}"
+    private_ip       = "${var.assign_private_ip == "true" ? cidrhost(lookup(var.network_cidrs,var.subnet_name), count.index+2) : ""}"
```
Comment (rohitkoul) [inline instances/k8smaster/main.tf]: this is static private ip right?

Comment (xinnong-wang) [reply instances/k8smaster/main.tf]: Yes, when assign_private_ip is set to true, this will be a static private ip.

File: k8s-oci.tf [hunk 8]
```diff
@@ -471,9 +470,15 @@ module "k8smaster-public-lb" {
...
+  hosts         = "${concat(module.instances-k8smaster-ad1.private_ips,module.instances-k8smaster-ad2.private_ips, module.instances-k8smaster-ad3.private_ips)}"
+  nginx_version = "${var.nginx_version}"
+}
+
 module "kubeconfig" {
   source                     = "./kubernetes/kubeconfig"
   api_server_private_key_pem = "${module.k8s-tls.api_server_private_key_pem}"
   api_server_cert_pem        = "${module.k8s-tls.api_server_cert_pem}"
-  k8s_master                 = "https://${module.k8smaster-public-lb.ip_addresses[0]}:443"
+  k8s_master                 = "{local.master_lb_address}"
```
Comment (jlamillan) [inline k8s-oci.tf]: There's still a typo here. `"{local.master_lb_address}"` is an invalid value. I am sure you meant `"${local.master_address}"` (note the missing `$`)

Beyond the typo, when `master_lb_enabled = false` and `control_plane_subnet_access = "public"`, we want the value of `k8s_master` to be based on `master_public_ip` or a `worker_public_ip`.

The purpose of the kubeconfig module is to generate the file `generated/kubeconfig` for the person running Terraform to use.  `https://127.0.0.1:6443` is not the right value when `master_lb_enabled = false` and `control_plane_subnet_access = "public"`.

File: k8s-oci.tf [hunk 9]
```diff
@@ -471,9 +470,15 @@ module "k8smaster-public-lb" {
   shape                     = "${var.k8sMasterLBShape}"
 }
 
+module "nginx-lb" {
```
Comment (jlamillan) [inline k8s-oci.tf]: Let's rename this module to `reverse-proxy`.

File: k8s-oci.tf [hunk 10]
```diff
@@ -471,9 +470,15 @@ module "k8smaster-public-lb" {
...
+  source        = "./network/loadbalancers/nginx"
```
Comment (jlamillan) [inline k8s-oci.tf]: Similarly, lets move / rename this to `./network/loadbalancers/software/reverse-proxy/k8smaster`

File: docs/input-variables.md [hunk 11]
```diff
@@ -21,6 +21,7 @@ name                                | default                 | description
 etcdShape                           | VM.Standard1.1          | OCI shape for etcd nodes
 k8sMasterShape                      | VM.Standard1.1          | OCI shape for k8s master(s)
 k8sWorkerShape                      | VM.Standard1.2          | OCI shape for k8s worker(s)
+master_oci_lb_enabled               | true                    | enable/disable the k8s master oci load balancer. true: use oci load balancer for k8s master . false: use nginx reverse proxy as a software load balancer for k8s masters.
```
Comment (jlamillan) [inline docs/input-variables.md]: "false: use nginx reverse proxy as a software load balancer for k8s masters" => "false: use software-based reverse proxy as a software load balancer for k8s masters"

File: network/loadbalancers/k8smaster/variables.tf (hunk 2 above)
Comment (jlamillan) [reply network/loadbalancers/k8smaster/variables.tf]: From the link"

> It is recommended for now to specify boolean values for variables as the strings "true" and "false", to avoid some caveats in the conversion process. A future version of Terraform will properly support boolean values and so relying on the current behavior could result in backwards-incompatibilities at that time.

Here's why:

```
grep master_oci_lb_enabled terraform.tfvars
master_oci_lb_enabled = "false"
```

```
terraform plan

Error: Error running plan: 2 error(s) occurred:

* module.k8smaster-public-lb.oci_load_balancer.lb-k8smaster: cannot parse "false" as an integer
* module.etcd-lb.oci_load_balancer.lb-etcd: cannot parse "false" as an integer
```
* module.etcd-lb.oci_load_balancer.lb-etcd: cannot parse "false" as an integer
* module.k8smaster-public-lb.oci_load_balancer.lb-k8smaster: cannot parse "false" as an integer

File: variables.tf [hunk 12]
```diff
@@ -215,42 +216,51 @@ variable "etcdLBShape" {
...
-  default = "true"
+  default     = true
 }
 
 variable "k8sMasterLBShape" {
   default = "100Mbps"
 }
 
+variable "master_oci_lb_enabled" {
+  description = "enable/disable the k8s master load balancer. true: use the k8s master load balancer ip. false: use a nginx reverse proxy for k8s masters."
```
[the remark this answers is not in the document]
Comment (jlamillan) [reply variables.tf]: From https://www.terraform.io/docs/configuration/variables.html#booleans:

> It is recommended for now to specify boolean values for variables as the strings "true" and "false", to avoid some caveats in the conversion process. A future version of Terraform will properly support boolean values and so relying on the current behavior could result in backwards-incompatibilities at that time.

Here's why:

```
grep master_oci_lb_enabled terraform.tfvars
master_oci_lb_enabled = "false"
```

```
terraform plan

Error: Error running plan: 2 error(s) occurred:

* module.k8smaster-public-lb.oci_load_balancer.lb-k8smaster: cannot parse "false" as an integer
* module.etcd-lb.oci_load_balancer.lb-etcd: cannot parse "false" as an integer
```

**Comment by jlamillan** [comment]
Xinnong,

I made a few minor changes, re-based, squashed, and pushed to a new branch. Could you [review](https://github.com/oracle/terraform-kubernetes-installer/commit/38990167f51731bf7378d66273f4abc147472ce8) and let me know what think. If it looks good, I can merge this tomorrow.

**Comment by xinnong-wang** [comment]
@jlamillan It looks fine to me.

**Comment by jlamillan** [comment]
Great work Xinnong! This is a nice addition. I merged this on the side. Closing this PR since your commit is now in the master branch.
